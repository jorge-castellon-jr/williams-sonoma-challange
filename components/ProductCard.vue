<template>
  <v-hover v-slot="{ hover }">
    <v-card v-if="content" class="product__card" hover>
      <v-img
        :height="content.thumbnail.height"
        :src="content.thumbnail.href"
        :alt="content.thumbnail.alt"
        :rel="content.thumbnail.rel"
      >
        <v-expand-transition>
          <div
            v-if="hover"
            class="justify-center text-center d-flex align-center transition-fast-in-fast-out orange darken-2 v-card--reveal display-3 white--text"
            style="height: 100%"
          >
            <span v-if="content.priceRange">
              ${{ convertDecimal(content.priceRange.selling.low) }} - ${{
                convertDecimal(content.priceRange.selling.high)
              }}
            </span>

            <span v-if="content.price">
              <div
                :class="
                  content.price.regular == content.price.selling
                    ? ''
                    : 'text-decoration-line-through'
                "
              >
                ${{ convertDecimal(content.price.regular) }}
              </div>
              <div v-if="content.price.regular != content.price.selling">
                ${{ convertDecimal(content.price.selling) }}
              </div>
            </span>
          </div>
        </v-expand-transition>
      </v-img>

      <v-card-title>
        {{ content.name }}
        <span class="caption">Tap/Hover to view Price</span>
      </v-card-title>

      <v-card-text class="align-self-end">
        <v-row align="center" class="mx-0">
          <v-rating
            :value="content.averageRating"
            color="amber"
            dense
            half-increments
            readonly
            size="14"
          ></v-rating>

          <div class="ml-4 grey--text">{{ displayReviews }}</div>
        </v-row>
      </v-card-text>
      <v-card-actions>
        <v-spacer />
        <v-btn :to="`/product/${content.id}`" color="orange darken-2" text>
          Learn More
        </v-btn>
      </v-card-actions>
    </v-card>
  </v-hover>
</template>

<script>
export default {
  name: 'ProductCard',
  props: {
    content: {
      type: Object,
      required: true,
    },
  },
  methods: {
    convertDecimal(value) {
      return value.toFixed(2)
    },
  },
  computed: {
    displayReviews() {
      return this.content.reviewCount > 0
        ? `${this.content.averageCount} (${this.content.reviewCount})`
        : 'No Reviews'
    },
  },
}
</script>

<style lang="scss" scoped>
.product {
  &__card {
    transition: box-shadow 0.3s ease;
    height: 100%;
    padding-bottom: 38px;
    position: relative;
    cursor: auto;
  }
}

.v-card {
  &__title {
    word-break: keep-all;
  }
  &__text {
    position: absolute;
    bottom: 0;
  }
}
</style>