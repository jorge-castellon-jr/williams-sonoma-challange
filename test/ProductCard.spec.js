import { createLocalVue, mount } from '@vue/test-utils'
import Vuetify from 'vuetify'

import ProductCard from '@/components/ProductCard.vue'

describe('Product Card Component', () => {
  const localVue = createLocalVue()
  let vuetify, nuxt

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  const mountProductCard = (options) => {
    return mount(ProductCard, {
      localVue,
      vuetify,
      ...options,
    })
  }

  it('is a Vue instance', () => {
    const wrapper = mountProductCard({
      propsData: {
        content: {
          id: 'test-id-1234',
          name: 'Bearaby Velvet Weighted Blanket - Silver Topaz',
          priceRange: {
            selling: {
              high: 299,
              low: 269,
            },
          },
          thumbnail: {
            size: 'm',
            meta: '',
            alt: '',
            rel: 'thumbnail',
            width: 363,
            href:
              'https://assets.weimgs.com/weimgs/rk/images/wcm/products/202046/0098/bearaby-velvet-weighted-blanket-silver-topaz-1-m.jpg',
            height: 363,
          },
          hero: {
            size: 'm',
            meta: '',
            alt: '',
            rel: 'hero',
            width: 363,
            href:
              'https://assets.weimgs.com/weimgs/rk/images/wcm/products/202046/0098/bearaby-velvet-weighted-blanket-silver-topaz-1-m.jpg',
            height: 363,
          },
          reviews: {
            recommendationCount: 0,
            likelihood: 0,
            reviewCount: 0,
            averageRating: 0,
          },
        },
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
