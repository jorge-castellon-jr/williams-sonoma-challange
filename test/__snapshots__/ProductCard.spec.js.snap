// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card Component is a Vue instance 1`] = `
"<div class=\\"product__card v-card v-card--hover v-sheet theme--light\\">
  <div aria-label=\\"\\" class=\\"v-image v-responsive theme--light\\" style=\\"height: 363px;\\" rel=\\"thumbnail\\">
    <transition-stub name=\\"fade-transition\\" mode=\\"in-out\\">
      <div class=\\"v-image__image v-image__image--preload v-image__image--cover\\" style=\\"background-position: center center;\\"></div>
    </transition-stub>
    <div class=\\"v-responsive__content\\">
      <transition-stub name=\\"expand-transition\\">
        <!---->
      </transition-stub>
    </div>
  </div>
  <div class=\\"v-card__title\\">
    Bearaby Velvet Weighted Blanket - Silver Topaz
    <span class=\\"caption\\">Tap to view Price</span></div>
  <div class=\\"v-card__text align-self-end\\">
    <div class=\\"row mx-0 align-center\\">
      <div class=\\"v-rating v-rating--readonly v-rating--dense\\"><button type=\\"button\\" tabindex=\\"-1\\" class=\\"v-icon notranslate v-icon--link mdi mdi-star-outline theme--light accent--text\\" style=\\"font-size: 14px;\\"></button><button type=\\"button\\" tabindex=\\"-1\\" class=\\"v-icon notranslate v-icon--link mdi mdi-star-outline theme--light accent--text\\" style=\\"font-size: 14px;\\"></button><button type=\\"button\\" tabindex=\\"-1\\" class=\\"v-icon notranslate v-icon--link mdi mdi-star-outline theme--light accent--text\\" style=\\"font-size: 14px;\\"></button><button type=\\"button\\" tabindex=\\"-1\\" class=\\"v-icon notranslate v-icon--link mdi mdi-star-outline theme--light accent--text\\" style=\\"font-size: 14px;\\"></button><button type=\\"button\\" tabindex=\\"-1\\" class=\\"v-icon notranslate v-icon--link mdi mdi-star-outline theme--light accent--text\\" style=\\"font-size: 14px;\\"></button></div>
      <div class=\\"ml-4 grey--text\\">No Reviews</div>
    </div>
  </div>
  <div class=\\"v-card__actions\\">
    <div class=\\"spacer\\"></div> <a class=\\"v-btn v-btn--flat v-btn--router v-btn--text theme--light v-size--default orange--text text--darken-2\\"><span class=\\"v-btn__content\\">
        Learn More
      </span></a>
  </div>
</div>"
`;
